### PR TITLE
Emit location update events to optional channel

### PR DESCRIPTION
**Pull request by mi-yu**
Addresses #2789 

Introduces a way for `InscriptionUpdater` to emit location update events on an in-process channel. Example usage can be found here: https://github.com/mi-yu/ord-custom-indexer-example

cc @casey

### CODE REVIEW

File: src/index.rs
Comment (casey) [inline src/index.rs:176]: Let's make this an enum, with a variant for new inscriptions, and a variant for old inscriptions that have move:

```rust
enum Event {
  InscriptionCreated {
    id: InscriptionId,
    location: Option<SatPoint>,
    sequence_number: u32,
    block: u32,
    charms: u16,
    parent: Option<InscriptionId>,
  }
  InscriptionMoved {
    id: InscriptionId,
    old_location: SatPoint,
    new_location: SatPoint,
    block: u32,
    sequence_number: u32,
  }
}
```

- Location can be an option when created, and be `None` if unbound.
- When an inscription moves, both old_location and new_location should be set, since unbound inscriptions don't move.
- I like having charms just be a u16, since no charms is indicated by `0`.

File: src/index.rs
Comment (casey) [inline src/index.rs:207]:
```suggestion
  event_sender: Option<tokio::sync::mpsc::Sender<LocationUpdateEvent>>,
```

File: src/index/updater.rs
Comment (casey) [inline src/index/updater.rs:546]: Can this be `self.location_update_sender.as_ref()`?

File: src/index/updater.rs
Comment (casey) [inline src/index/updater.rs:636]: Can the inscription updater be constructed with the event sender, so we don't have to pass it through?

**Review by casey — CHANGES_REQUESTED** [review]
Just a quick initial review. I'm going to be traveling on and off until the 22nd of December, so I might not be able to do much review until then.

File: src/index.rs
Comment (mi-yu) [reply src/index.rs]: thanks for the suggestion, makes sense

**Comment by mi-yu** [comment]
@casey bumping this

File: .gitignore
Comment (casey) [inline .gitignore:13]:
```suggestion
/.vscode
```

File: src/index.rs
Comment (casey) [inline src/index.rs:195]: Let's make the name just `Event`, in case we want to add events that aren't related to inscriptions.

```suggestion
#[derive(Debug, Clone)]
pub enum Event {
```

File: src/index.rs
Comment (casey) [inline src/index.rs:209]:
```suggestion
    id: InscriptionId,
    location: Option<SatPoint>,
    sequence_number: u32,
    block_height: u32,
    charms: u16,
    parent_id: Option<InscriptionId>,
```

File: src/index.rs
Comment (casey) [inline src/index.rs:221]:
```suggestion
    id: InscriptionId,
    old_location: SatPoint,
    new_location: SatPoint,
    sequence_number: u32,
    block_height: u32,
```

File: src/index.rs
Comment (casey) [inline src/index.rs:417]:
```suggestion
  pub fn set_event_sender(
```

File: src/index.rs
Comment (casey) [inline src/index.rs:424]:
```suggestion
  ) {
    self.event_sender = Some(sender);
  }
```

File: src/index.rs
Comment (casey) [reply src/index.rs:195]: Let's also move this into a submodule, `src/index/event.rs`

File: src/index/updater.rs
Comment (casey) [inline src/index/updater.rs:41]: The updater has the index, so we shouldn't need to store the event sender separately.

```suggestion
```

File: src/index/updater.rs
Comment (casey) [inline src/index/updater.rs:54]:
```suggestion
```

File: src/index.rs
Comment (casey) [inline src/index.rs:240]: Fields should be sorted.

File: src/index/updater.rs
Comment (casey) [inline src/index/updater.rs:449]: Fields should be sorted.

File: src/index/updater.rs
Comment (casey) [reply src/index/updater.rs:41]: Fields should be sorted.

File: src/index/updater/inscription_updater.rs
Comment (casey) [inline src/index/updater/inscription_updater.rs:488]:
```suggestion
            location: (!unbound).then_some(new_satpoint),
```

**Review by casey — CHANGES_REQUESTED** [review]
Nice, mostly looks good!

This needs a test, probably in `index.rs`, that events are emitted when an inscription is created, and when an inscription is moved. Let me know if you need some pointers on how to write it.

**Comment by mi-yu** [comment]
Thanks for the review @casey, I've added a test checking that the `Event::InscriptionCreated` is sent through the channel but not sure the best way to test `Event::InscriptionMoved`, do you have recommendations there?

File: .gitignore
Comment (casey) [inline .gitignore:13]: These entries should be sorted.

File: src/index/event.rs
Comment (casey) [inline src/index/event.rs:8]:
```suggestion
```

File: src/index/event.rs
Comment (casey) [inline src/index/event.rs:10]:
```suggestion
    inscription_id: InscriptionId,
```

File: src/index/event.rs
Comment (casey) [inline src/index/event.rs:11]: These should be sorted.

File: src/index/event.rs
Comment (casey) [inline src/index/event.rs:18]:
```suggestion
    inscription_id: InscriptionId,
```

Comment (casey) [inline src/index/event.rs:18]: These should be sorted

File: src/index.rs
Comment (casey) [inline src/index.rs:383]: Instead of having a setter, let's have two `open` functions, `Index::open` and `Index::open_with_event_sender`, `Index::open` can call `Index::open_with_event_sender` with None.

File: src/index.rs
Comment (casey) [inline src/index.rs:5828]: We should also test the `InscriptionTransferred` event.

File: src/index/event.rs
Comment (casey) [inline src/index/event.rs:4]:
```suggestion
```

File: src/index/updater/inscription_updater.rs
Comment (casey) [inline src/index/updater/inscription_updater.rs:44]: This should be:
```suggestion
  pub(super) event_sender: Option<&'a Sender<Event>>,
```

**Review by casey — CHANGES_REQUESTED** [review]
Added some comments, take a look. If you can allow me to push changes to this branch, I can make minor fixes myself, which will help this get in faster.

**Comment by raphjaph** [comment]
@casey I think this is ready for review

**Review by casey — CHANGES_REQUESTED** [review]
Looks good! A couple of things:

- Let's change `Event::InscriptionMoved` to `Event::InscriptionTransferred`
- There are conflicts that need to be fixed

I would fix them myself, but this PR doesn't allow changes to be made by maintainers.

**Comment by casey** [comment]
@mi-yu Can you please enable maintainers to push to this branch? I often want to make small changes to a branch before merging, and without being able to make this changes myself, each one is a full round trip of requesting a change, waiting for it to be made, and then reviewing it again.

You need to click this button:

Screenshot 2024-02-17 at 1 13 01 PM

**Comment by mi-yu** [comment]
@casey seems I don't have that option since the fork is owned by the magicoss org, let me fork to my personal GitHub and reopen the PR. Thanks!

**Comment by mi-yu** [comment]
New one here @casey https://github.com/ordinals/ord/pull/3137, closing this one

**Pull request closed without merging**
